src/tests-number-predicate.cc

```diff
@@ -0,0 +1,83 @@
+//
+// Created by Lukáš Holík on 06.12.2022.
+//
+
+#include "../3rdparty/catch.hpp"
+
+#include <mata/nfa.hh>
+
+using namespace Mata::Util;
+using namespace Mata::Nfa;
+
+TEST_CASE("Mata::Util::NumPredicate") {
+    NumPredicate<State> p;
+    // to test switching between watching and not watching
+    std::vector<bool> vals = {true, false, false, true, true};
+    for (auto val: vals) {
+        if (val) p.watch_elements();
+        else p.dont_watch_elements();
+
+//        SECTION("basic functionality: add, remove, access, constructor, size, get_elements") {
+            std::vector<State> v = {1, 2, 3, 4, 5};
+            p.add(v);
+            auto elems = p.get_elements();
+            CHECK(elems == v);
+            CHECK(p.size() == 5);
+            p.remove({2, 4});
+            CHECK(OrdVector<State>(p) == OrdVector<State>({1,3,5}));
+            v = {1, 3, 5};
+            elems = p.get_elements();
+            CHECK(elems == v);
+            CHECK(p.size() == 3);
+            CHECK(p.size() == 3);
+            p.add({1, 3, 5});
+            elems = p.get_elements();
+            CHECK(elems == v);
+            p.remove({1, 2, 3, 4, 5});
+            elems = p.get_elements();
+            v = {};
+            CHECK(elems == v);
+            for (State q = 0; q < 10; q++) CHECK(!p[q]);
+            CHECK(p.size() == 0);
+//        }
+
+//        SECTION("iterator") {
+            p.add({1, 2, 3, 4, 5});
+            State i = 1;
+            for (auto q: p) {
+                CHECK(q == i);
+                i++;
+            }
+            CHECK(i == 6);
+            p.clear();
+//        }
+
+//        SECTION("accessing stuff outside current domain") {
+            CHECK(!p[100]);
+            p.add(100);
+            CHECK(p[100]);
+            CHECK(!p[99]);
+            CHECK(!p[101]);
+            CHECK(p.size() == 1);
+//        }
+
+//        SECTION("complement") {
+        p = {2,4};
+        p.complement(5);
+        CHECK(OrdVector<State>(p) == OrdVector<State>({0,1,3}));
+        p.complement(6);
+        CHECK(OrdVector<State>(p) == OrdVector<State>({2,4,5}));
+
+        p = {2,4,8};
+        p.complement(6);
+        CHECK(OrdVector<State>(p) == OrdVector<State>({0,1,3,5}));
+
+        p.complement(6);
+        CHECK(OrdVector<State>(p) == OrdVector<State>({2,4}));
+
+        p = {};
+//        }
```

Comment:
Could you remove the commented out lines or uncomment them in case they are meant to be there?

Reply:
I actually wanted to ask someone how the sections work. 
If they are not commented out, they get exetuted only once, even though they are in the for-loop. What I am doing wrong?

Reply:
Sorry, I did not notice they are in a for loop. I have read about this problem once, even if I did not need to try the recommended solution myself. This is caused by the underlying implementation of section macro, which is just an if statement checking for a correct section name in order to execute the current section.

Hopefully, the recommended approach in the linked [issue](https://github.com/catchorg/Catch2/issues/993) should fix that.

Reply:
nice, that actually works

Reply:
Great to hear that the fix is still relevant. We will need to keep this in mind for the future.